### ISSUE TITLE
250702 Machine Learning Basic_안희원

### ISSUE THREAD

**Issue by notone0818** [AIBootcamp14/wiki-rider#7]
## 📅 학습 날짜  
2025-07-02

## 📌 키워드 및 요약 정리
_**# Unsupervised Learning**, **# Clustering**, **# Dimensionality Reduction**, **# Manifold Hypothesis**, **# High Dimensional Data**_

- **Unsupervised Learning**은 label이 없는 데이터에서 숨겨진 구조나 패턴을 찾아내는 학습방식으로 하위 분야로는 크게 *(1) Clustering*과 *(2) Dimensionality Reduction*이 있다. 
- 특징 : 🔹 label x 🔹 Denoising 🔹 feedback x 🔹 데이터 내 latent structure 발견
- **Clustering**는 유사한 데이터들을 군집화 하는 작업으로 k-Means, DBSCAN 등와 같은 기법이 있다.
- **Dimensionality Reduction**은 고차원 데이터를 저차원 표현으로 투영하여 핵심 구조를 유지하는 방식으로 PCA, t-SNE 등와 같은 기법이 있다.

### 1️⃣ Machine Learning Overview ⭐

머신러닝은 크게 나누어 보면 **(1) supervised learning**, **(2) unsupervised learning** (3) reinforcement learning으로 나눌 수 있다. *(semi-supervised learning은 생략 하겠다.)*

| 종류        | 특징 1              | 특징 2        | 특징 3                         |
|------------------|---------------------|----------------|---------------------------------|
| supervised learning | 레이블 된 데이터     | 직접 피드백     | 출력 및 미래 예측              |
| unsupervised learning   | 레이블 없이 feature만 존재| 피드백 없음     | 데이터에서 숨겨진 구조 찾기     |
| reinforcement learning    | 결정 과정           | 보상 시스템     | 연속된 행동에서 학습           |

*[Table 1] Comparative Summary of Supervised, Unsupervised, and Reinforcement Learning*

---
#### 🔹 **Supervised Learning** : target & feautre 간의 관계 학습

- **Classification** : Logistic Regression, SVM, Decision Tree, Random Forest ...
- **Regression** : Linear Regression, Ridge Regression, Lasso Regression...
- [여기로 가서 자세히 보세요_feat.예은님](https://github.com/AIBootcamp14/wiki-rider/issues/5)

#### 🔹 **Unsupervised Learning** : label 없이 데이터의 구조 및 패턴 추출
- **Clustering** : k-means, DBSCAN ...
- **Dimensionality Reduction** : PCA, t-SNE ...
---
### 2️⃣ Unsupervised Learning ⭐
**(DEF)**
label이 없는 데이터에서 데이터 간 유사성, latent structure 등을 학습하는 방식으로 입력만 존재하고 label이 없기 때문에 피드백 없이 스스로 pattern을 찾아낸다.

**(PURPOSE)**
- Clustering ; K-means, DBSCAN ...
- Dimensionality / Visualization ; PCA, t-SNE ...
- Denoising (or) Data Preprocessing ; Autoencoder...
- Latent Variable 추정
---
#### ✅ **Clustering**
**(DEF)**
label이 없이 데이터로부터 pattern을 파악해 여러 개의 cluster로 나누는 방식이다.
- **K-means** : 유클리드 거리 기반, 구형 cluster에 강함
- DBSCAN : 밀도 기반 clustering, 노이즈 및 이상치에 강함

**(VS) Classification**
- **Classification**은 주어진 feature을 input으로 사용해, class를 추론하고 모델의 결과와 label을 비교해 학습    
- **Clustering**은 주어진 sample을 어떻게 나누어야 가장 유사한 sample끼리 묶이게 될지 data 간 관계를 비교해 학습

  ❗ 즉, Label의 유/무로 정답이 주어졌냐? 아니냐? 

  ❗예를 들어, image를 cat/dog으로 나누는 task를 classification과 clustering으로 진행해 보았다고 하자. **Classification**는 분류하고자 하는 class가 설정한 cat/dog 대로 나뉜다. (정확도는 이번에 따지지 않는다고 하자.) 다만, **Clustering**의 경우 k(2라 가정)만 주어지기에 2가지 군집으로 나뉘기만 하지 이게 cat/dog인지는 까보기 전까지 모른다.

---
#### ✅ **Dimensionality Reduction**
**(DEF)** 고차원 데이터를 의미 있는 저차원 표현으로 축소하는 방식으로 핵심 정보는 보존하면서 noise 제거, 시각화 등을 수행한다.
- **PCA** : 선형 변환 기반, 분산 최대 보존
- **t-SNE** : 비선형 거리 기반 축소, 시각화에 특화
---
#### ✅ **Code Example** 
※ chatGPT : PCA, k-means, t-SNE를 활용하여 가장 간단한 예제 하나 달라고 한 결과 코드 입니다. 참고 바랍니다.

❗하단에 sklearn 공식 문서로 가는 링크 첨부해 두었습니다.

🔹Dataset = Iris

🔹Skill 
1. PCA : 4D -> 2D [ [[바로가기 : sklearn_PCA]]](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)
2. K-means : 3개의 cluster 군집화 [[[바로가기 : sklearn_k_means]]](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html)
3. t-SNE : 시각화 표현 [[[바로가기 : sklearn_t-SNE]]](https://scikit-learn.org/stable/modules/manifold.html)

``` python
from sklearn.datasets import load_iris
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt

# 1. 데이터 로딩 (4차원 feature)
iris = load_iris()
X = iris.data

# 2. PCA: 고차원 데이터를 2D로 축소 (선형)
X_pca = PCA(n_components=2).fit_transform(X)

# 3. K-means: PCA 결과에 기반한 군집화 (비지도)
kmeans = KMeans(n_clusters=3, random_state=42)
clusters = kmeans.fit_predict(X_pca)

# 4. t-SNE: 2차원에서의 비선형 시각화 투영
X_tsne = TSNE(n_components=2, perplexity=30, random_state=42).fit_transform(X_pca)

# 5. 시각화: 군집 결과 기반 시각화
plt.figure(figsize=(8, 6))
plt.scatter(X_tsne[:, 0], X_tsne[:, 1], c=clusters, cmap='viridis', s=50)
plt.title("t-SNE Visualization of PCA + k-Means", fontsize=14)
plt.xlabel("t-SNE dim 1")
plt.ylabel("t-SNE dim 2")
plt.colorbar(label="Cluster ID")
plt.grid(True, linestyle='--', alpha=0.5)
plt.show()
```
---
### 3️⃣ Data Labeling Cost 💸 
**Data Labeling(annotation)** 작업은 상당한 비용과 시간을 필요로 하다. Unsupervised Learning은 이 작업이 필요 없으므로 많은 데이터를 쉽게 구할 수 있다.

⚠️ **주의할 점**

Unsupervised Learning은 단순히 **라벨링 비용을 줄이기 위한 목적**으로 고안된 것이 아니다. Supervised Learning과 Unsupervised Learning은 근본적으로 해결하고자 하는 문제 정의 자체가 다르다.

➕ Unsupervised 방법으로 Clustering을 적용했으나, 성능이 지나치게 낮아 직접 약 10,000건의 라벨링을 수행해 semi-supervised 방법으로 진행한 바가 있다. 🫠
 
이런 접근은 특히 NLP 분야에서 흔히 발생한다고도 알려져있다.

---
### 4️⃣ Manifold Hypothesis 💌
**(DEF)**
대부분의 고차원 (real-world) data는 실제로는 낮은 차원의 곡면(manifold)에 존재한다.

**(KEY)** *Embedding*은 고차원 데이터에서 의미 있는 저차원 표현을 추출하는 작업이다. 이 과정은 데이터가 존재하는 manifold의 구조를 보존하려는 목표를 가진다.⭐

**(EX)** 한국어 텍스트 데이터의 경우 [KoNLPy, KoBERT]등의 도구를 통해 벡터화 하는 것을 의미한다. 이렇게 얻어진 embedding은 자연어가 분포하는 manifold을 학습할 수 있도록 한다.

---
### 5️⃣ High Dimensional Data
**(DEF)**
Feature의 수가 많아 차원이 높은 데이터를 의미하며, 현실 세계의 데이터는 일반적으로 High-Dimensional 특성을 갖는다.

**(EX)** 
- 🖼 Image : M * N * 3(RGB)
- 📄 Text : Embedding하면 단어당 약 300~1024차원

**(PROBLEMS)**
1. Computational Complexity
2. Sparsity of Data
3. Loss of Intuition / Interpretablility

---

## ❓ 어려웠던 부분
- Clustering 평가 지표(예: 실루엣 점수 등)의 의미와 한계
- Clustering 결과의 품질을 어떻게 객관적으로 판단할 수 있을지에 대한 고민

## 💬 질문 또는 토론 유도 
- t-SNE와 UMAP의 차이점 및 장단점이란? [[umap_paper_link]](https://arxiv.org/abs/1802.03426)
- TTS(Text-to-Speech)와 같은 경우 문제가 음성이 깨지거나 여러 음성이 겹치는 경우, 복원이나 추출을 하려고 unsupervised learning이 활용되는가?

**Comment by Young-SeokOh** [comment]
희원 선생님..감동이네요. 정리를 너무 잘하셨네요. 나중에 브런치 스토리 작가를 하셔도 되겠습니다😄 지금처럼 계속해서 성실히 하시다보면 좋은 결과가 있으리라 생각합니다 :)

필요한 페이퍼도 본인이 다 찾으셨고, 따로 피드백 드리지 않아도 선생님 스스로 해결하실 수 있을 것으로 보입니다. 정말 수고 많으셨습니다.

**Comment by LM-GIJUN** [comment]
안녕하세요 희원님!
저도 오영석 멘토님과 같은 마음입니다...
진짜 감동이네요........ 예은님의 어시스트도 좋았습니다 👍

"Clustering 결과의 품질을 어떻게 객관적으로 판단할 수 있을지에 대한 고민" 이 고민은 저도 공부 할 때 많이 했던 고민인데 저는 시각화로 많이 해소하였습니다. 시각화 하나로는 많은 한계가 있어 다양한 지표들을 활용하긴 해야 하지만 저는 전체적인 흐름 파악을 위해 시각화를 통해 많이 해소하긴 했습니다! (대충한거 티내는 중)

TTS에서 노이즈, 음성이 겹치는 것 등을 라벨링하는 것은 엄청난 노가다 작업이기 때문에 비지도 학습으로 패턴 학습시키는 것이 유리해보이네요!
최근에는 음성의 경우 자기지도학습이 대세라고 하네요 (관련 논문 : https://arxiv.org/abs/2204.03421)

희원님 학습하시면서 정리까지 하시느라 정말 고생많으셨습니다!
